content/blog/dplyr-1-1-0-joins/index.Rmd

```diff
@@ -0,0 +1,237 @@
+
+To actually fix this issue, we'll need to expand our join specification to include another condition.
+Let's zoom in to just 2021:
+
+```{r}
+filter(faulty, company == "B", year == 2021)
+```
+
+We want to retain the match with RStudio, but not with Posit (because the name hasn't changed yet).
+One way to express this is by using the `year` and `since` columns to state that you only want a match if the transaction `year` occurred *after* a name change:
+
+```{r}
+# `year[i] >= since`?
+2021 >= 2009
+2021 >= 2022
+```
+
+Because `join_by()` accepts expressions, we can express this inequality directly inside the join specification:
+
+```{r}
+join_by(company == id, year >= since)
+```
+
+```{r}
+transactions |>
+  inner_join(companies, join_by(company == id, year >= since))
+```
+
+This eliminated the 2021 match to Posit, as expected!
+This type of join is known as an *inequality join*, i.e. it involves at least one join expression containing one of the following inequality conditions: `>=`, `>`, `<=`, or `<`.
+
+However, we still have 2 matches corresponding to the 2023 year.
+In this case, we only wanted the match to Posit.
+We can understand why we are still getting multiple matches here by running the same row-by-row analysis as before:
+
+```{r}
+# `year[i] >= since`? Both are true!
+2023 >= 2009
+2023 >= 2022
+```
+
+To remove the last problematic match of the 2023 transaction to the RStudio name, we'll need to refine our join specification one more time.
+
+## Rolling joins
+
+Inequality conditions like `year >= since` are powerful, but since the condition is only bounded on one side it is common for them to return a large number of matches.
+Since multiple matches are the typical case with inequality joins, we don't get a warning like with the equality join, but we clearly still haven't gotten the join right.
+As a reminder, here are where we still have too many matches:
+
+```{r}
+transactions |>
+  inner_join(companies, join_by(company == id, year >= since)) |>
+  filter(company == "B", year == 2023)
+```
+
+We need a way to filter down the matches returned from `year >= since` to only the most recent name change.
+In other words, we prefer the Posit match over the RStudio match because 2022 is *closer* to the transaction year of 2023 than 2009 is.
+We can express this in `join_by()` by using a helper named `closest()`.
+
+```{r}
+transactions |>
+  inner_join(companies, join_by(company == id, closest(year >= since)))
+```
+
+`closest(year >= since)` finds all of the matches in `since` for a particular `year`, and then filters them down to only the closest match to that `year`.
+This is known as a *rolling join*, because in this case it *rolls* the most recent name change forward to match up with the transaction.
+Rolling joins were popularized by data.table, and are related to `ASOF` joins supported by some SQL flavors.
```

Comment:
```suggestion
Rolling joins were popularized by data.table, and are related to "as of" joins supported by some SQL flavors.
```
?

Reply:
I always see it as `ASOF` https://questdb.io/docs/reference/sql/join/#:~:text=ASOF%20JOIN,timestamp%20in%20the%20first%20table.

I always get frustrated when google searching for this because `as of join` doesn't pull anything up, it has to be `asof join` 😆

Reply:
Then how about `` `ASOF` ("as of") `` because I tried to figure out what the acronym stood for.